protoc-gen-go/generator/generator.go

```diff
@@ -2508,6 +2509,97 @@ func (g *Generator) generateMessage(message *Descriptor) {
 	g.addInitf("%s.RegisterType((*%s)(nil), %q)", g.Pkg["proto"], ccTypeName, fullName)
 }
 
+func unescape(s string) string {
+	// NB: Sadly, we can't use strconv.Unquote because protoc will escape both
+	// single and double quotes, but strconv.Unquote only allows one or the
+	// other (based on actual surrounding quotes of its input argument).
+	s, err := doUnescape(s)
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+	return s
+}
+
+func doUnescape(s string) (string, error) {
```

Comment:
This seems much more complicated than necessary.

See this version:
https://play.golang.org/p/DHt_EPXYyt

Reply:
I did not test mine, but you get the idea.

Reply:
Also, according to https://en.wikipedia.org/wiki/Escape_sequences_in_C#Table_of_escape_sequences,
the octal sequence may be be less than 3 octal digits.

Also, should we support `\e` or `\Uhhhhhhhh` or `\uhhhh`? I don't know if protoc even generates those.

Reply:
Also, what happens if the input string is non-ASCII. Does protoc ever generate that?

Reply:
> I don't know if protoc even generates those.

The precise implementation of protoc currently is that it will never emit unicode escapes or '\e'`. In fact, it only emits octal ecapes, and they are not variable length (always 3 chars).
https://github.com/google/protobuf/blob/master/src/google/protobuf/stubs/strutil.cc#L590

I was worried about coupling this too tightly to the current implementation (don't want it to break if protoc impl changes), which is why I added code for hex escapes. But since the field's type is `bytes`, not `string`, it seems highly unlikely that it should be trying to support unicode escapes (and, by extension, UTF-8 encoding).

> Also, what happens if the input string is non-ASCII. Does protoc ever generate that?

Currently, protoc does not generate such a string. But, if it did, the bytes of non-ascii characters (as long as they don't include `\`) would be conveyed in the value, which seems correct given the field's type is `bytes`.

Reply:
Sounds good. We can revisit for non-ASCII and unicode runes if it's an issue in the future.